File: LeetCode/easy/Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, string s, vector<string>& v) {
        if (s == "") s = s + to_string(root->val);
        else s = s + "->" + to_string(root->val);
        if(root->left == NULL && root->right == NULL) v.push_back(s);
        if(root->left != NULL) dfs(root->left, s, v);
        if(root->right != NULL) dfs(root->right, s, v);
    }
    
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> v;
        if (root == NULL) return v;
        dfs(root, "", v);
        return v;
    }
};

File: Lintcode/480_Binary Tree Paths/Binary Tree Paths.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root the root of the binary tree
     * @return all root-to-leaf paths
     */
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> list = new ArrayList<String>();
        if (root != null) addPath(root, String.valueOf(root.val), list);
        return list;        
    }
    
    public void addPath(TreeNode node, String path, List<String> list) {
        if (node == null) return;
        if (node.left == null && node.right == null) list.add(path);
        if (node.left != null) addPath(node.left, path + "->" + String.valueOf(node.left.val), list);
        if (node.right != null) addPath(node.right, path + "->" + String.valueOf(node.right.val), list);
    }
    
}